### PR TITLE
Etcd KVClient implementation.

### LINKED ISSUE #1144: etcd support
Many Kubernetes admins already know how to deal with etcd. It would be good if you could switch out consul for etcd to reduce the cognitive load on admins.

### PR DESCRIPTION
**Pull request by tomwilkie**
Should allow us to store the ring in etcd.

Fixes #1144

### CODE REVIEW

File: pkg/ring/kvstore.go [hunk 1]
```diff
@@ -44,66 +34,41 @@ func (cfg *KVConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
...
+		return consul.NewConsulClient(cfg.Consul, codec)
 	case "inmemory":
 		// If we use the in-memory store, make sure everyone gets the same instance
 		// within the same process.
 		inmemoryStoreInit.Do(func() {
-			inmemoryStore = NewInMemoryKVClient(codec)
+			inmemoryStore = consul.NewInMemoryKVClient(codec)
 		})
 		return inmemoryStore, nil
+	case "etcd":
```
Comment (cstyan) [inline pkg/ring/kvstore.go]: nit, can we have `inmemory` always be the last case before default?

File: pkg/ring/kvstore/consul/consul_client.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package ring
+package consul
```
Comment (cstyan) [inline pkg/ring/kvstore/consul/consul_client.go]: :+1:

File: pkg/ring/kvstore/consul/consul_client_mock.go [hunk 3]
```diff
@@ -18,7 +19,7 @@ type mockKV struct {
 }
 
 // NewInMemoryKVClient makes a new mock consul client.
-func NewInMemoryKVClient(codec Codec) KVClient {
+func NewInMemoryKVClient(codec kvstore.Codec) kvstore.KVClient {
```
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/ring/kvstore/consul/consul_client_mock.go]: Or just `NewInMemoryClient`?

**Comment by bboreham** [comment]
Given this code, is it possible to move a running Cortex instance from Consul to etcd ?

I suspect it would need something like a way to configure ingesters to update both destinations, then when that's up, configure distributors to read from the new one, then turn off the old one.

File: pkg/ingester/lifecycle_test.go [hunk 4]
```diff
@@ -195,6 +198,21 @@ func TestIngesterBadTransfer(t *testing.T) {
 	require.Equal(t, ring.PENDING, ing.lifecycler.GetState())
 }
 
+func numTokens(c kvstore.KVClient, name string) int {
```
Comment (cstyan) [inline pkg/ingester/lifecycle_test.go]: this doesn't seem to be used anywhere?

File: pkg/ring/kvstore.go [hunk 5]
```diff
@@ -1,36 +1,27 @@
...
-// such as CAS and Watch which take callbacks.  It also deals with serialisation
-// by having an instance factory passed in to methods and deserialising into that.
-type KVClient interface {
-	CAS(ctx context.Context, key string, f CASCallback) error
-	WatchKey(ctx context.Context, key string, f func(interface{}) bool)
-	WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool)
-	Get(ctx context.Context, key string) (interface{}, error)
-	PutBytes(ctx context.Context, key string, buf []byte) error
-}
+var inmemoryStore kvstore.KVClient
```
Comment (cstyan) [inline pkg/ring/kvstore.go]: this doesn't seem to be used anymore, we previously assigned to it in `NewKVStore`

File: pkg/ring/kvstore/consul/consul_client_mock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/ring/kvstore/consul/consul_client_mock.go]: bump, still makes more sense for this to be `client_mock.go` imo.

File: pkg/ring/kvstore/etcd/etcd.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/cortexproject/cortex/pkg/ring/kvstore"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	"github.com/etcd-io/etcd/clientv3"
+	"github.com/go-kit/kit/log/level"
+)
+
+// Config for a new etcd.Client.
+type Config struct {
+	Endpoints   []string      `yaml:"endpoints"`
```
Comment (cstyan) [inline pkg/ring/kvstore/etcd/etcd.go]: hmm okay, `Endpoints` is each etcd leader node I guess?

File: pkg/ring/kvstore/kvstore.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+)
+
+// KVClient is a high-level client for Consul, that exposes operations
```
Comment (cstyan) [inline pkg/ring/kvstore/kvstore.go]: high-level client for various key value stores, such as Consul or etcd, that exposes operations

File: pkg/ring/kvstore/kvstore.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	return snappy.Encode(nil, bytes), nil
+}
+
+type prefixedKVClient struct {
+	prefix string
+	client KVClient
+}
+
+// PrefixClient takes a ConsulClient and forces a prefix on all its operations.
```
Comment (cstyan) [inline pkg/ring/kvstore/kvstore.go]: takes a  KVClient

File: pkg/ring/kvstore/kvstore.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+// WatchPrefix watches a prefix. For a prefix client it appends the prefix argument to the clients prefix.
+func (c *prefixedKVClient) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	c.client.WatchPrefix(ctx, fmt.Sprintf("%s%s", c.prefix, prefix), func(k string, i interface{}) bool {
+		return f(strings.TrimPrefix(k, c.prefix), i)
+	})
+}
+
+// PutBytes writes bytes to Consul.
+func (c *prefixedKVClient) PutBytes(ctx context.Context, key string, buf []byte) error {
```
Comment (cstyan) [inline pkg/ring/kvstore/kvstore.go]: writes bytes to the KVStore

File: pkg/ring/lifecycler.go [hunk 10]
```diff
@@ -144,7 +145,7 @@ func NewLifecycler(cfg LifecyclerConfig, flushTransferer FlushTransferer, name s
 	if port == 0 {
 		port = *cfg.ListenPort
 	}
-	codec := ProtoCodec{Factory: ProtoDescFactory}
+	codec := kvstore.ProtoCodec{Factory: ProtoDescFactory}
```
Comment (cstyan) [inline pkg/ring/lifecycler.go]: why not just use `ring.GetCodec()` here as well?

File: pkg/ring/kvstore/consul/consul_client_mock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/ring/kvstore/consul/consul_client_mock.go]: > bump, still makes more sense for this to be client_mock.go imo.

This isn't a mock of the KVClient interface, but a mock of the package-local `kv` interface wrapped in a `consulClient`.  It's really to test the `consulClient` implementation, and its unfortunate we've used it elsewhere.

We should follow up this a PR to tidy this up and introduce a mock of the `KVClient` interface, but for now I think this is in the right place.

File: pkg/ring/kvstore/etcd/etcd.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+		}
+
+		buf, err := c.codec.Encode(intermediate)
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "error serialising value", "key", key, "err", err)
+			continue
+		}
+
+		_, err = c.cli.Txn(ctx).
+			If(clientv3.Compare(clientv3.Version(key), "=", revision)).
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/ring/kvstore/etcd/etcd.go]: Tell me about it...

File: pkg/ring/kvstore.go (hunk 5 above)
Comment (tomwilkie) [reply pkg/ring/kvstore.go]: Thats a bug; the inmemorystore needs to be a singleton for the single process stuff to work.

Fixed.

File: pkg/ring/kvstore/etcd/etcd.go (hunk 6 above)
Comment (tomwilkie) [reply pkg/ring/kvstore/etcd/etcd.go]: I think you can point it at _any_ etcd node.

**Comment by tomwilkie** [comment]
> Given this code, is it possible to move a running Cortex instance from Consul to etcd ?

As it stands, it is not.  Agreed, needs a migration path.  Would you mind if this was done in another PR?  Right now we need this for the HA Prometheus pairs code, which easily overloads Consul.

**Comment by tomwilkie** [comment]
@gouthamve have another look - I move a bunch of code around to make the package names and imports much cleaner.

File: pkg/ring/kvstore/consul/consul_client_mock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/ring/kvstore/consul/consul_client_mock.go]: Have renamed it to `NewInMemoryClient`

File: pkg/ring/kv/etcd/etcd.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+			level.Error(util.Logger).Log("msg", "error CASing", "key", key, "err", err)
+			if !retry {
+				return err
+			}
+			continue
+		}
+
+		if intermediate == nil {
+			panic("Callback must instantiate value!")
+		}
```
Comment (tomwilkie) [inline pkg/ring/kv/etcd/etcd.go]: This needs to match the consul client for HA tracker, which sometimes doesn't want to finish the CAS.

**Comment by tomwilkie** [comment]
> where is the vendor coreos/etcd coming from? libraries used by etcd-io clientv3?

Have fixed this, everything should be coming from `go.etcd.io/etcd` now.

File: pkg/ring/kv/etcd/etcd.go (hunk 12 above)
Comment (tomwilkie) [reply pkg/ring/kv/etcd/etcd.go]: Fixed.

File: pkg/ring/kv/client.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	// We need Consul flags to not have the ring prefix to maintain compatibility.
+	// This needs to be fixed in the future (1.0 release maybe?) when we normalize flags.
+	// At the moment we have consul.<flag-name>, and ring.store, going forward it would
+	// be easier to have everything under ring, so ring.consul.<flag-name>
+	cfg.Consul.RegisterFlags(f, prefix)
+	cfg.Etcd.RegisterFlagsWithPrefix(f, prefix)
+	if prefix == "" {
+		prefix = "ring."
+	}
+	f.StringVar(&cfg.Prefix, prefix+"prefix", "", "The prefix for the keys in the store. Should end with a /.")
```
Comment (bboreham) [inline pkg/ring/kv/client.go]: Default here changes from `"collectors/"` - breaking change?

**Pull request merged**